### ISSUE TITLE
MPS memory leak

### ISSUE THREAD

**Issue by SalmanMohammadi** [pytorch/torchtune#2473]
Originally raised in Discord by @maximegmd. Running the default Qwen2.5-0.5B config with `device=mps` results in a memory leak before OOM, on torch version 2.6.1 and on nightly.

![Image](https://github.com/user-attachments/assets/9db07aef-77e9-45f7-8cfd-9aa3edbf82ee)

Training logs:
```bash
(.venv) combuter:~/projects/torchtune $ tune run full_finetune_single_device --config qwen2_5/0.5B_full_single_device tokenizer.path=./target/Qwen/Qwen2.5-0.5B-Instruct/vocab.json tokenizer.merges_file=./target/Qwen/Qwen2.5-0.5B-Instruct/merges.txt checkpointer.checkpoint_dir=./target/Qwen/Qwen2.5-0.5B-Instruct/ tokenizer.max_seq_len=512 device=mps
import error: No module named 'triton'
W0310 10:58:36.962000 44162 torch/distributed/elastic/multiprocessing/redirects.py:29] NOTE: Redirects are currently not supported in Windows or MacOs.
Running FullFinetuneRecipeSingleDevice with resolved config:

batch_size: 2
checkpointer:
  _component_: torchtune.training.FullModelHFCheckpointer
  checkpoint_dir: ./target/Qwen/Qwen2.5-0.5B-Instruct/
  checkpoint_files:
  - model.safetensors
  model_type: QWEN2
  output_dir: /tmp/torchtune/qwen2_5_0_5B/full_single_device
  recipe_checkpoint: null
clip_grad_norm: null
compile: false
dataset:
  _component_: torchtune.datasets.alpaca_cleaned_dataset
  packed: false
device: mps
dtype: bf16
enable_activation_checkpointing: false
enable_activation_offloading: false
epochs: 1
gradient_accumulation_steps: 1
log_every_n_steps: 1
log_peak_memory_stats: true
loss:
  _component_: torchtune.modules.loss.CEWithChunkedOutputLoss
max_steps_per_epoch: null
metric_logger:
  _component_: torchtune.training.metric_logging.StdoutLogger
  log_dir: /tmp/torchtune/qwen2_5_0_5B/full_single_device/logs
model:
  _component_: torchtune.models.qwen2_5.qwen2_5_0_5b
optimizer:
  _component_: torch.optim.AdamW
  fused: true
  lr: 2.0e-05
optimizer_in_bwd: false
output_dir: /tmp/torchtune/qwen2_5_0_5B/full_single_device
profiler:
  _component_: torchtune.training.setup_torch_profiler
  active_steps: 2
  cpu: true
  cuda: true
  enabled: false
  num_cycles: 1
  output_dir: /tmp/torchtune/qwen2_5_0_5B/full_single_device/profiling_outputs
  profile_memory: false
  record_shapes: true
  wait_steps: 5
  warmup_steps: 3
  with_flops: false
  with_stack: false
resume_from_checkpoint: false
seed: null
shuffle: true
tokenizer:
  _component_: torchtune.models.qwen2_5.qwen2_5_tokenizer
  max_seq_len: 512
  merges_file: ./target/Qwen/Qwen2.5-0.5B-Instruct/merges.txt
  path: ./target/Qwen/Qwen2.5-0.5B-Instruct/vocab.json

Setting manual seed to local seed 1249317817. Local seed is seed + rank = 1249317817 + 0
Model is initialized with precision torch.bfloat16.
Memory stats after model init:
        CPU peak memory active: 0.96 GiB
        CPU peak memory alloc: 1.06 GiB
Tokenizer is initialized from file.
Optimizer is initialized.
Loss is initialized.
No learning rate scheduler configured. Using constant learning rate.
 Profiling disabled.
 Profiler config after instantiation: {'enabled': False}
1|1| ...       | 1/25880 [00:00<6:25:31,  1.12it/s]Step 1 | loss:2.3462696075439453 lr:2e-05 tokens_per_second_per_gpu:383.07843017578125 peak_memory_active:2.867405891418457 peak_memory_alloc:6.09918212890625 
1|2| ...       | 2/25880 [00:01<5:22:35,  1.34it/s]Step 2 | loss:2.0462706089019775 lr:2e-05 tokens_per_second_per_gpu:390.3084716796875 peak_memory_active:2.867405891418457 peak_memory_alloc:7.3840789794921875 
1|3| ...       | 3/25880 [00:02<4:45:06,  1.51it/s]Step 3 | loss:1.882745623588562 lr:2e-05 tokens_per_second_per_gpu:272.8753662109375 peak_memory_active:2.867403984069824 peak_memory_alloc:7.747100830078125 
1|4| ...       | 4/25880 [00:02<4:37:07,  1.56it/s]Step 4 | loss:1.4600402116775513 lr:2e-05 tokens_per_second_per_gpu:413.9355163574219 peak_memory_active:2.867405891418457 peak_memory_alloc:8.000732421875 
1|5| ...       | 5/25880 [00:03<4:36:42,  1.56it/s]Step 5 | loss:1.491969108581543 lr:2e-05 tokens_per_second_per_gpu:450.4296569824219 peak_memory_active:2.867405891418457 peak_memory_alloc:8.254364013671875 
1|6| ...       | 6/25880 [00:03<4:34:15,  1.57it/s]Step 6 | loss:1.1335465908050537 lr:2e-05 tokens_per_second_per_gpu:581.3117065429688 peak_memory_active:2.867405891418457 peak_memory_alloc:8.50799560546875 
1|7| ...       | 7/25880 [00:04<5:16:04,  1.36it/s]Step 7 | loss:1.3159548044204712 lr:2e-05 tokens_per_second_per_gpu:763.9815673828125 peak_memory_active:2.867424726486206 peak_memory_alloc:9.865158081054688 
1|8| ...       | 8/25880 [00:05<5:07:40,  1.40it/s]Step 8 | loss:1.333193302154541 lr:2e-05 tokens_per_second_per_gpu:301.0128173828125 peak_memory_active:2.867403984069824 peak_memory_alloc:10.196929931640625 
1|9| ...       | 9/25880 [00:06<4:44:54,  1.51it/s]Step 9 | loss:1.3530874252319336 lr:2e-05 tokens_per_second_per_gpu:652.633056640625 peak_memory_active:2.867405891418457 peak_memory_alloc:10.157867431640625 
1|10| ...      | 10/25880 [00:07<5:15:05,  1.37it/s]Step 10 | loss:0.6320281624794006 lr:2e-05 tokens_per_second_per_gpu:385.632568359375 peak_memory_active:2.867424726486206 peak_memory_alloc:10.411636352539062 
1|11| ...      | 11/25880 [00:08<6:39:31,  1.08it/s]Step 11 | loss:1.0807266235351562 lr:2e-05 tokens_per_second_per_gpu:477.3770446777344 peak_memory_active:2.867424726486206 peak_memory_alloc:12.747299194335938 
1|12| ...      | 12/25880 [00:09<7:00:38,  1.02it/s]Step 12 | loss:1.0825313329696655 lr:2e-05 tokens_per_second_per_gpu:483.7933044433594 peak_memory_active:2.867424726486206 peak_memory_alloc:12.946258544921875 
1|13| ...      | 13/25880 [00:10<6:08:01,  1.17it/s]Step 13 | loss:1.3548258543014526 lr:2e-05 tokens_per_second_per_gpu:407.2728576660156 peak_memory_active:2.867403984069824 peak_memory_alloc:13.285842895507812 
1|14| ...      | 14/25880 [00:10<6:01:09,  1.19it/s]Step 14 | loss:1.3036937713623047 lr:2e-05 tokens_per_second_per_gpu:286.06634521484375 peak_memory_active:2.867405891418457 peak_memory_alloc:13.521896362304688 
1|15| ...      | 15/25880 [00:11<6:29:13,  1.11it/s]Step 15 | loss:1.150942087173462 lr:2e-05 tokens_per_second_per_gpu:554.5146484375 peak_memory_active:2.867424726486206 peak_memory_alloc:13.775543212890625 
1|16| ...      | 16/25880 [00:12<6:20:25,  1.13it/s]Step 16 | loss:1.3549091815948486 lr:2e-05 tokens_per_second_per_gpu:431.3746337890625 peak_memory_active:2.867424726486206 peak_memory_alloc:13.990127563476562 
1|17| ...      | 17/25880 [00:13<6:33:31,  1.10it/s]Step 17 | loss:1.2903493642807007 lr:2e-05 tokens_per_second_per_gpu:789.15673828125 peak_memory_active:2.867424726486206 peak_memory_alloc:13.990127563476562 
1|18| ...      | 18/25880 [00:14<6:50:41,  1.05it/s]Step 18 | loss:1.5478922128677368 lr:2e-05 tokens_per_second_per_gpu:498.7770080566406 peak_memory_active:2.867424726486206 peak_memory_alloc:14.2437744140625 
1|19| ...      | 19/25880 [00:15<6:52:16,  1.05it/s]Step 19 | loss:1.58563232421875 lr:2e-05 tokens_per_second_per_gpu:542.18896484375 peak_memory_active:2.867424726486206 peak_memory_alloc:14.495468139648438 
1|20| ...      | 20/25880 [00:16<6:35:59,  1.09it/s]Step 20 | loss:1.2682877779006958 lr:2e-05 tokens_per_second_per_gpu:502.465087890625 peak_memory_active:2.867424726486206 peak_memory_alloc:14.81549072265625 
1|21| ...      | 21/25880 [00:17<6:21:29,  1.13it/s]Step 21 | loss:1.058565616607666 lr:2e-05 tokens_per_second_per_gpu:501.17657470703125 peak_memory_active:2.867424726486206 peak_memory_alloc:15.069107055664062 
1|22| ...      | 22/25880 [00:18<6:32:04,  1.10it/s]Step 22 | loss:1.126811146736145 lr:2e-05 tokens_per_second_per_gpu:495.6554260253906 peak_memory_active:2.867424726486206 peak_memory_alloc:15.334457397460938 
1|23| ...      | 23/25880 [00:18<5:59:56,  1.20it/s]Step 23 | loss:1.0712127685546875 lr:2e-05 tokens_per_second_per_gpu:502.4743957519531 peak_memory_active:2.867405891418457 peak_memory_alloc:15.293441772460938 
1|24| ...      | 24/25880 [00:20<6:21:48,  1.13it/s]Step 24 | loss:1.3631954193115234 lr:2e-05 tokens_per_second_per_gpu:711.341064453125 peak_memory_active:2.867424726486206 peak_memory_alloc:15.56658935546875 
1|25| ...      | 25/25880 [00:20<5:54:42,  1.21it/s]Step 25 | loss:1.180518627166748 lr:2e-05 tokens_per_second_per_gpu:498.5926208496094 peak_memory_active:2.867405891418457 peak_memory_alloc:15.869155883789062 
1|26| ...      | 26/25880 [00:21<5:34:27,  1.29it/s]Step 26 | loss:1.2827892303466797 lr:2e-05 tokens_per_second_per_gpu:437.0075988769531 peak_memory_active:2.867405891418457 peak_memory_alloc:16.122787475585938 
1|27| ...      | 27/25880 [00:22<5:47:02,  1.24it/s]Step 27 | loss:1.8088457584381104 lr:2e-05 tokens_per_second_per_gpu:541.8067626953125 peak_memory_active:2.867424726486206 peak_memory_alloc:16.376434326171875 
```

Raising this here instead of core as I'm not sure what the root cause is. 
cc @malfet

**Comment by felipemello1** [comment]
@SalmanMohammadi , does os.environ["PYTORCH_CUDA_ALLOC_CONF"] = "expandable_segments:True" help

**Comment by SalmanMohammadi** [comment]
> PYTORCH_CUDA_ALLOC_CONF

I haven't tried this but I'm not sure if this will make a difference we're not using cuda in this case?

> dataset.packed=True 

Unfortunately this still resulted in the same behaviour.

**Comment by musabgultekin** [comment]
I reproduced the issue and can confirm the peak_memory_alloc is increasing.

I tried profiling. I did `profiler.profile_memory=True` but I got `Segmentation fault: 11` when I enabled . But that was the only log printed out.

Then I added `faulthandler.enable()` to see what is going on.

I saw that training crashes with "Segmentation fault" error at this line
`logits = self._model(**batch)`

But that was it. So im stuck there. Wasn't able to run memory profiler for MPS.

It *only* happens `profiler.profile_memory=True`

**Comment by ebsmothers** [comment]
@musabgultekin I'm not sure about the memory leak, but at least regarding memory profiling we rely on torch.profiler (see [here](https://github.com/pytorch/torchtune/blob/main/torchtune/training/_profiler.py#L363)). However, seems like for MPS there is a separate torch.mps.profiler module (see [e.g.](https://pytorch.org/docs/stable/generated/torch.mps.profiler.profile.html)). So that may explain why profiling isn't working for you

**Comment by minhquoc0712** [comment]
Any update on this?

**Comment by SalmanMohammadi** [comment]
Hi @minhquoc0712. I've potentially tracked this issue down to a memory leak in SDPA and raised it [here](https://github.com/pytorch/pytorch/issues/152344).

For now, one solution you could try is setting `dataset.packed=True` - I found that the memory usage stabilised after some steps.

**Comment by minhquoc0712** [comment]
Hi @SalmanMohammadi Are you still encountering the memory problem with the pytorch fixed mentioned in your issue?

**Comment by SalmanMohammadi** [comment]
@minhquoc0712 unfortunately so. I raised https://github.com/issues/created?issue=pytorch%7Cpytorch%7C152550 for another issue I was encountering.
